### PR TITLE
util/types:Handle out of range value when `load data`

### PR DESCRIPTION
**Pull request by louishust**
When insert an overflow string like "1343545435346432587475"
to MySQL int type column, tidb will convert to max int64
instead of max int.

### CODE REVIEW

**Comment by louishust** [comment]
solve bug #3024

File: util/types/datum_test.go [hunk 1]
```diff
@@ -131,6 +133,15 @@ func testDatumToInt64(c *C, val interface{}, expect int64) {
 	c.Assert(b, Equals, expect)
 }
 
+func testDatumToInt64WithTargetType(c *C, val interface{}, tp byte, expectErr Checker, expectVal int64) {
```
Comment (XuHuaiyu) [inline util/types/datum_test.go]: you can add this test it in util/types/convert_test.go, 
thus `ToInt64WithTp `, `testDatumToInt64WithTargetType` and `testDatumToInt64WithTargetType` can be removed.

**Comment by XuHuaiyu** [comment]
PTAL @zimulala

**Comment by shenli** [comment]
@louishust Should we distinguish between loading data and insert statement?

mysql> create table t (a int, b int not null);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (143, 1343545435346432587475);
ERROR 1264 (22003): Out of range value for column 'b' at row 1
mysql>

**Comment by louishust** [comment]
@shenli 

```
mysql> set sql_mode="";
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values(1, "1343545435346432587475");
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> select * from t1;
+------+------------+
| a    | b          |
+------+------------+
|  143 | 2147483647 |
|    1 | 2147483647 |
+------+------------+
2 rows in set (0.00 sec)
```

I think the sql_mode is modified at load data command internal in MySQL.
I will check that in MySQL later.

**Comment by louishust** [comment]
@shenli 

I checked MySQL code and find the following code in sql/sql_yacc.yy:

```
load:
.............
        if (lex->local_file && lex->duplicates == DUP_ERROR)
              lex->set_ignore(true);
```
So if load data with local keyword, the load will treat truncate error as warnings.

And tidb only support load data local infile, so should always ignore the error, 
and tidb already handled this situation in executor/write.go:

```
func (e *LoadDataInfo) insertData(cols []string) {
	for i := 0; i < len(e.row); i++ {
		if i >= len(cols) {
			e.row[i].SetString("")
			continue
		}
		e.row[i].SetString(cols[i])
	}
	row, err := e.insertVal.fillRowData(e.columns, e.row, true)
```
The fillRowData's third parameter is **true**.

**Comment by shenli** [comment]
@zimulala PTAL

**Comment by hanfei1991** [comment]
LGTM

**Comment by zimulala** [comment]
If using the statement of `load data infile` behavior is like `sql_mode` is empty behavior, then we can directly set `sql_mode` value to empty when handling the statement of `load data infile`. Because the [function](Please add more tests to verify your claim) considers the value of `sql_mode`.
And please add more tests to verify your claim.

**Comment by zimulala** [comment]
@louishust

**Comment by louishust** [comment]
@zimulala Ignore the sql_mode, only LOCAL keyword matters.

File: util/types/datum.go [hunk 2]
```diff
@@ -1327,6 +1327,8 @@ func (d *Datum) toSignedInteger(sc *variable.StatementContext, tp byte) (int64,
 	case KindString, KindBytes:
 		iVal, err := StrToInt(sc, d.GetString())
 		if err != nil {
+			// adjust the overflow value
+			iVal, _ = ConvertIntToInt(iVal, lowerBound, upperBound, tp)
```
Comment (zimulala) [inline util/types/datum.go]: I'm sorry,  I forgot `StrToInt` implementation details.
You can consider it like the [logical](https://github.com/pingcap/tidb/pull/3508/files#diff-b67a235702396fe5cc8ec3d67962697eR1366).@louishust

**Comment by louishust** [comment]
@zimulala modified the logic

**Pull request merged**